CAM-11692: add joining gateway assertion

ingorichtsmeier opened the pull request:
Add support to assert joining gateways

ingorichtsmeier commented:
implementation of feature request https://jira.camunda.com/browse/CAM-11692

ingorichtsmeier commented:
An alternative to adding a new assertion for joining gateways would be to change the implementation of `isWaitingAt` and check the activity instance tree here than just the active activities. 

The original problem is, that `isWaitingAt`didn't work for joining gateways, as they are no active activities.

@tmetzke , what do you think?

mboskamp commented:
Hi @ingorichtsmeier,

thank you for the contribution. I have forwarded this PR and the related CAM issue to @ThorbenLindhauer for further prioritization.

Cheers,
Miklas

ingorichtsmeier commented:
Hi @ThorbenLindhauer, 

I changed the API back to `isWaitingAt()` and all current tests are green on my computer. I will add further tests to check your comments:

> With an ActivityInstance-based implementation, we would begin matching activities that are not leaves in the hierarchy (the subprocess in Example i above). From my perspective that is ok, what do you think?

It is OK, but I will a test to clarify.
> In order to make an ActivityInstance-based implementation work with async continuations, we must also look at TransitionInstances (async continuations are represented as these objects in the tree, see ActivityInstance#getChildTransitionInstances).

More tests required here.

Stay tuned. I would like to get this published in the next release. What's the deadline for this?

ThorbenLindhauer commented:
Hey Ingo, the code freeze for Assert 6.0.0 (with Camunda BPM 7.13) is the 13th of May, i.e. it must be merged by then. It wouldn't be a problem to build a 6.1.0 quickly after that, but I would also prefer to have this in 6.0.0.

ingorichtsmeier commented:
Hi @ThorbenLindhauer, 

I adjusted the code to handle async continuation and subprocesses. First adding tests and changed the implementation to run the tests successfully.

As the transition instances are quite complicated to handle I added a logger to the project to be able to see some output in the console running tests.

Could you please have a look at my changes and merge them if they are OK?

Thank you, Ingo

ThorbenLindhauer commented on core/src/main/java/org/camunda/bpm/engine/test/assertions/AssertionsLogger.java:
```diff
@@ -0,0 +1,42 @@
+  public static final AssertionsLogger INSTANCE = BaseLogger.createLogger(
+      AssertionsLogger.class, "BPM_ASSERT", "org.camunda.bpm.engine.test.assertions", "00");
+  
+  public void collectTransitionInstances(ActivityInstance activityInstance) {
+    logDebug("001", "collecting transition instances for activity instance '{}' with ID '{}'", 
+        activityInstance.getActivityName(),
+        activityInstance.getActivityId());
+  }
```
The log message is a bit misleading. The activity instance is uniquely identified by its ID. I would log that instead of the activity name or change the content of the log message such that it is clear that the name is logged. The same applies to the transition instance logging.

ingorichtsmeier commented:
Hi @ThorbenLindhauer, 

> We now slightly change the behavior of `#isWaitingAtExactly` when a task is located in a subprocess. A user will now have to provide the subprocess IDs as well. Do you think it should be that way or will users rather see this as a chore?

Thats difficult to predict. I assume that the usage of `isWaitingAtExactly()` in combination with nested subprocesses is quite rare. So I vote for keeping it as it is right now. 

Disable the activity instance tree for `isWaitingAtExactly()` will make the implementation more complex. And it's unintuitive that you cannot check for joining gateways with `isWaitingAtExactly()`.

I've adjusted the log messages as you proposed.

Ready to merge?

Cheers, Ingo

ThorbenLindhauer commented:
Hey Ingo,

I merged your contribution via https://github.com/camunda/camunda-bpm-assert/commit/5758eca7bbe264735fa22edd8e679a93c55bab75. I made some more changes to the logging and refactored the tests a bit to better fit the given-when-then structure that we have defined in our [testing best practices](https://github.com/camunda/camunda-bpm-dev-docs/blob/master/development/Testing-Best-Practices-Java.md).

Thanks for your contribution!

Cheers,
Thorben
